### PR TITLE
Add support for specifying the GC/Wii system language per-game

### PR DESCRIPTION
**Pull request by iwubcode**
There are some Japanese Wii games that require you to set your wii system language to "Japanese" in order for them to run.  Being a native English speaker, I'd prefer to keep my system language to "English" and only change it to "Japanese" when playing these specific games.  This pull request allows me to specify the wii language in the specific game ini.

I've tested on my Japanese copy of Disaster Day Of Crisis.  Captain Rainbow didn't have this issue.  Still waiting to receive two more Japanese games in the mail so any testing would be helpful here (I'm thinking of updating the game specific inis in the future).  I've also noticed from the wiki that [Tales of Graces](https://wiki.dolphin-emu.org/index.php?title=Tales_of_Graces) exhibits this issue.

EDIT:  This is my first github merge request, so if I've done something wrong, please bear with me :)

### CODE REVIEW

**Comment by JosJuice** [comment]
A better solution for that specific problem would be to make the language override option work for Wii games. (It only works for GC games right now.) But this solution isn't bad either. There might be other reasons for wanting to force a game to a specific language, making this useful. Do we allow forcing a specific language for GC games in game INIs right now? If not, we should have that too to match.

(This isn't a code review, I'm only reviewing the general approach)

**Comment by iwubcode** [comment]
> Do we allow forcing a specific language for GC games in game INIs right now? If not, we should have that too to match.

Fairly new to this code but it doesn't look like it.  

> A better solution for that specific problem would be to make the language override option work for Wii games. (It only works for GC games right now.)

Not aware of this, how do you do a language override?  Admittedly, the language stuff in the code base confused me (there's a lot of it), so I might have missed something.

**Comment by JosJuice** [comment]
> My understanding is the "SelectedLanguage" in Core is the Gamecube language?

Yes.

> Not aware of this, how do you do a language override? Admittedly, the language stuff in the code base confused me (there's a lot of it), so I might have missed something.

The option is currently called "Override Language on NTSC Games" in the GUI, and it's implemented here: https://github.com/dolphin-emu/dolphin/blob/d5ca153c26a9eae2084a3ff8d4a8d5cd1acf56ac/Source/Core/Core/BootManager.cpp#L353

It works like this: Dolphin checks if the selected language isn't intended to be used with the current game's region, and if so, it sets another language for the current emulation session. This is simpler to implement for GameCube than for Wii.

On GameCube, NTSC-U and NTSC-J both only have the valid language 0 (though it means English for the former and Japanese for the latter) and PAL has valid languages 0-5, so all we have to do is set the language to 0 if NTSC is used and the language isn't already 0.

On Wii, English is now 1 instead of 0, NTSC-U supports more languages, there's an entirely new Korean region with an associated Korean language, and there are also Chinese language options (but are they a separate region?). I don't think anyone knows exactly what's allowed for each Wii region, so if we are to implement language override for Wii games, we would need to find out about that first.

**Comment by JosJuice** [comment]
Sorry, I got it backwards. What I described will happen when the override option is off, not on.

**Comment by iwubcode** [comment]
Thanks for the explanation.  The automatic override feels a little hacky but maybe that's just me.  I'd almost feel like an alternative would be to create at least one more PR (if not more) that:

- Adds the ability to set the language for individual GC games
- Sets the override of the language to English for Baiten Kaitos and any other known gamecube games that have that issue
- Removes the override option from code / UI

**Comment by JosJuice** [comment]
I think it's less hacky, because it means we don't have to do anything on a per-game basis. Besides, the list of allowed languages per region is well-defined by Nintendo, not just something hacky we came up with. So please don't get rid of it or add overrides to specific game INIs. Adding the ability to set GC languages in game INIs sounds good to me, though.

**Comment by iwubcode** [comment]
Fair enough!  Would you like that to be in this pull request (and I update the title) or a different one?

**Comment by JosJuice** [comment]
Putting it in this pull request makes the most sense to me (but you can make a separate one if you want to).

**Comment by iwubcode** [comment]
@JosJuice  - done.  I wasn't sure how to test this, I did a quick regression check with the only GameCube game I have ripped: Soul Calibur II.

I noticed an issue with the "Movie" language and how Wii is not coming from the movie itself.  I'll probably save that for another future PR.

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by iwubcode** [comment]
Attempted to do a rebase / merge master to my fork.  I don't have to do these at work, so it's a bit unfamiliar, but I think it looks ok!

**Comment by leoetlino** [comment]
Nitpicking, but the commit titles shouldn't really have "- " in them (they're not list items) nor "feedback...:" as it isn't useful info to leave in the commit history forever.

**Comment by iwubcode** [comment]
@leoetlino - sorry, this is the style I use at my day job!  I think I can fix this with a rebase + reword?  Might as well do a squash as well I guess?  I can just use the title text as my commit.

We can't do rebases at work (not without using some separate remotes) so this is all new territory to me.

**Comment by iwubcode** [comment]
Fixed

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by Parlane** [comment]
@dolphin-emu-bot rebuild

**Comment by iwubcode** [comment]
Just curious, anything else I need to do to get this merged?

**Comment by Parlane** [comment]
@dolphin-emu-bot rebuild

**Comment by Parlane** [comment]
@JosJuice Could you please give this a last look over. You are more familiar with this area.

**Comment by JosJuice** [comment]
I'm not completely sure how the SYSCONF language handling works. @leoetlino should be better at it.

File: Source/Core/Core/BootManager.cpp [hunk 1]
```diff
@@ -300,6 +305,9 @@ bool BootCore(const std::string& _rFilename)
         g_wiimote_sources[WIIMOTE_BALANCE_BOARD] = source;
         WiimoteReal::ChangeWiimoteSource(WIIMOTE_BALANCE_BOARD, source);
       }
+
+      IniFile::Section* sysconf_section = game_ini.GetOrCreateSection("Sysconf");
```
Comment (JosJuice) [inline Source/Core/Core/BootManager.cpp]: Do we already have a section called "Sysconf", or is that new for this PR? If we don't already have one, would to be more appropriate to use an existing section?

Comment (leoetlino) [inline Source/Core/Core/BootManager.cpp]: We already have one in Dolphin.ini, but not in game INIs.

Comment (JosJuice) [inline Source/Core/Core/BootManager.cpp]: We should probably use the [Wii] section then. It already contains SYSCONF settings like widescreen, and "Wii" means a lot more for a user than "Sysconf".

Comment (leoetlino) [inline Source/Core/Core/BootManager.cpp]: That's true. @iwubcode, if you change it to use the Wii section, don't forget to change the WiiLanguage name to Language.

Comment (iwubcode) [inline Source/Core/Core/BootManager.cpp]: Sure, I think that is a good change to make.  One question though - don't you think we should have name parity between the two types of inis?  Not necessarily in this PR but maybe another to update things to match?

Comment (leoetlino) [inline Source/Core/Core/BootManager.cpp]: Well, the other names already don't match, so I don't think that's an issue.

Comment (JosJuice) [inline Source/Core/Core/BootManager.cpp]: Having consistent names would of course be nice, but the backwards compatibility issues sound very unfun to deal with. I don't think just expecting users to re-set many of their settings would be acceptable. But either way, it shouldn't be done in this PR.

**Comment by iwubcode** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/Core/BootManager.cpp (hunk 1 above)
Comment (JosJuice) [inline Source/Core/Core/BootManager.cpp]: Hmm, I wonder if we should rename the GC setting too... I don't think we have a GameCube section to put it in, but maybe we can call it something like GameCubeLanguage? "Selected" is rather redundant, and we should make it clear that this affects GameCube games, not Wii games or the GUI.

Comment (iwubcode) [inline Source/Core/Core/BootManager.cpp]: I thought that too but wouldn't it make sense to make another PR that changes all the settings to be under the "Gamecube" label?  What other settings are there?

Comment (JosJuice) [inline Source/Core/Core/BootManager.cpp]: Renaming any setting will lead to setting files being incompatible unless we implement some sort of migration. It might not be worth it.

There aren't really many settings that apply to the GameCube but not to the Wii. The only thing I can think of that for sure doesn't exist on a Wii is the SP1 slot. Memory cards can also be seen as a GameCube thing, although they technically work just as fine on a Wii - it's just that Wii games almost never make use of them. So even if we didn't have to worry about compatibility, I'm not sure if there's much we could put in a GameCube section.

Comment (iwubcode) [inline Source/Core/Core/BootManager.cpp]: @JosJuice - agreed.  I'm happy to do whatever but will say that it was confusing seeing "SelectedLanguage" in the code too because I didn't really realize till later that that was referring to the Gamecube language.  If we change the setting, I'd like to change the code references too.

Comment (JosJuice) [inline Source/Core/Core/BootManager.cpp]: Yes, it would be nice to change it in the code too. Feel free to do it in either this PR or a separate PR, but you should probably do it as a separate commit if you do it in this PR.

Comment (iwubcode) [inline Source/Core/Core/BootManager.cpp]: Thanks, I'll probably save that for a follow-up PR.  It seems there are a lot of 'cleanup' PRs and so that will probably fit rather nicely.

**Comment by iwubcode** [comment]
@JosJuice / @leoetlino  - any other concerns?

**Comment by JosJuice** [comment]
If @leoetlino thinks the SYSCONF handling is fine, I don't think there's anything left to do other than renaming the name of the GC setting in the INIs.

**Comment by leoetlino** [comment]
LGTM. It's not really possible to mess up the SYSCONF now that writing to it is handled by ConfigManager and only it.

**Comment by JosJuice** [comment]
Note that the INI setting rename needs to be done in this PR regardless of when it's renamed in the code, since users are going to start putting the old name in their INIs otherwise.

File: Source/Core/Core/BootManager.cpp [hunk 2]
```diff
@@ -252,6 +256,7 @@ bool BootCore(const std::string& _rFilename)
     core_section->Get("HLE_BS2", &StartUp.bHLE_BS2, StartUp.bHLE_BS2);
     core_section->Get("ProgressiveScan", &StartUp.bProgressive, StartUp.bProgressive);
     core_section->Get("PAL60", &StartUp.bPAL60, StartUp.bPAL60);
+    core_section->Get("GamecubeLanguage", &StartUp.SelectedLanguage, StartUp.SelectedLanguage);
```
Comment (JosJuice) [inline Source/Core/Core/BootManager.cpp]: It's supposed to be GameCube with a capital C.

File: Source/Core/Core/BootManager.cpp [hunk 3]
```diff
@@ -300,6 +305,9 @@ bool BootCore(const std::string& _rFilename)
         g_wiimote_sources[WIIMOTE_BALANCE_BOARD] = source;
         WiimoteReal::ChangeWiimoteSource(WIIMOTE_BALANCE_BOARD, source);
       }
+
+      IniFile::Section* sysconf_section = game_ini.GetOrCreateSection("Wii");
```
Comment (JosJuice) [inline Source/Core/Core/BootManager.cpp]: Sorry, I missed this before, but the variable name should be `wii_section` (since the section isn't called [Sysconf] anymore)

**Comment by psennermann** [comment]
How do I set the Wii system language per game? I mean, what do I have to write in the game.ini?
Thanks

**Comment by iwubcode** [comment]
@psennermann - here's an example of how you'd set your language to Japanese:

[Wii]
Language = 0

**Comment by psennermann** [comment]
works perfectly, thanks

**Comment by MayImilae** [comment]
@iwubcode Were you aware that this doesn't allow setting a GameCube game to Japanese? Apparently, the GameCube uses a VI bit to say "this is NTSC sure, but it's Japanese!) and the ForceNTSCJ option needs to be used for Japanese on the GameCube.

https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/Core/HW/VideoInterface.cpp#L184

**Comment by JosJuice** [comment]
@MayImilae In NTSC-J games, setting the language to 0 means Japanese. It only means English if you're in NTSC-U and PAL games. The NTSC-J video bit isn't related to languages as far as I know.

**Comment by MayImilae** [comment]
@iwubcode Ok! We did some more poking around, and on the GameCube 0 is both Japanese and English, and it changes depending on the game. However, NTSC-J GameCubes send that "this is NTSC sure, but it's Japanese!", and many japanese games will become half japanese half english if they don't see the NTSC-J flag. That's what ForceNTSC-J is for. Hopefully this craziness will be fixed soon, but for now, an NTSC-J gameINI option would be nice to deal with this.

**Comment by JosJuice** [comment]
> so for error free Japanese gameplay, you have to set it to English (0, for NTSC), and then turn on ForceNTSC-J to option sends the NTSC-J flag.

No, Dolphin automatically turns on the NTSC-J flag when booting Japanese games or the Japanese IPL, so the Force NTSC-J setting is unnecessary for getting things to run correctly unless you're booting DOL/ELF files directly.

**Comment by MayImilae** [comment]
@JosJuice What if you didn't bother to install a Japanese IPL? Just someone downloaded Dolphin, and ran a Japanese game?

**Comment by JosJuice** [comment]
Dolphin turns on the NTSC-J flag automatically in that case too.

**Comment by MayImilae** [comment]
Then why is ForceNTSCJ even in the UI? >_<

**Comment by JosJuice** [comment]
To clarify for others who are reading this (I was talking with May and HdkR on IRC): That auto detection was in old 2010 code. The current auto detection doesn't have the same problems as far as I know. (Though we still can't detect NTSC-J for DOL/ELF files, like I said before)
